### PR TITLE
Adding Custom Alphabet Encoder Class

### PR DESCRIPTION
**Pull request by marmegh**
Adding class to allow encoding given a custom alphabet sequence. Testing scenarios include, encode and decode produce consistent results for a given alphabet, encode produces expected strings for given inputs and alphabets, and unsupported input throws the correct error.

### CODE REVIEW

File: src/Marvin/CustomAlphabetEncoder.cs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        }
+
+        /// <summary>
+        /// Encode an unsigned integer (a checksum) in a given character set.
+        /// </summary>
+        /// <param name="data">The unsigned integer to encode.</param>
+        /// <returns>
+        /// The unsigned integer encoded using the character set with which this class was instantiated.
+        /// </returns>
+        public string Encode (uint data)
```
Comment (eddynaka) [inline src/Marvin/CustomAlphabetEncoder.cs:79]: >  [](http://example.com/codeflow?start=28&length=1)

nit: remove the white space

File: src/Tests.Marvin/CustomAlphabetEncoderTests.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            Action action = () => testEncoder.Encode(data);
+            Assert.ThrowsException<ArgumentOutOfRangeException>(action);
+        }
+
+        /// <summary>
+        /// Compare values of a byte array instead of references of a byte array.
+        /// </summary>
+        /// <param name="expected">The expected byte array content.</param>
+        /// <param name="actual">The byte array content being tested.</param>
+        public void AssertByteArraysAreEqual(byte[] expected, byte[] actual)
```
Comment (eddynaka) [inline src/Tests.Marvin/CustomAlphabetEncoderTests.cs:76]: >public [](http://example.com/codeflow?start=8&length=6)

change to private, since this won't be called by others

File: src/Tests.Marvin/CustomAlphabetEncoderTests.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        {
+            Assert.AreEqual(expected.Length, actual.Length);
+            
+            for(int i = 0; i < expected.Length; i++)
+            {
+                Assert.AreEqual(expected[i], actual[i]);
```
Comment (eddynaka) [inline src/Tests.Marvin/CustomAlphabetEncoderTests.cs:82]: >Assert.AreEqual [](http://example.com/codeflow?start=16&length=15)

we are using fluentAssertions, with that, you will be able to do something like:
actual.Should().Be(expected)

Comment (eddynaka) [reply src/Tests.Marvin/CustomAlphabetEncoderTests.cs:82]: when you do that, you will be able to remove the first assert (length assertion)

File: src/Tests.Marvin/CustomAlphabetEncoderTests.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Security.Utilities
+{
+    using System;
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+    [TestClass]
+    public class CustomAlphabetEncoderTests
```
Comment (eddynaka) [inline src/Tests.Marvin/CustomAlphabetEncoderTests.cs:10]: >CustomAlphabetEncoderTests [](http://example.com/codeflow?start=17&length=26)

you could add the following tests:
- input is null to cover the argumentnullexception for encode and decode
- a small alphabet (abc) and pass abc and a few other strings to validate

File: src/Marvin/CustomAlphabetEncoder.cs [hunk 5]
```diff
@@ -0,0 +1,119 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Security.Utilities
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Text;
```
Comment (eddynaka) [inline src/Marvin/CustomAlphabetEncoder.cs:9]: we normally add the usings outside the namespace

File: src/Tests.Marvin/CustomAlphabetEncoderTests.cs (hunk 3 above)
Comment (eddynaka) [reply src/Tests.Marvin/CustomAlphabetEncoderTests.cs]: if we are using should.beequivalent, i think u dont need to compare each index value. Can you verify?

Comment (marmegh) [reply src/Tests.Marvin/CustomAlphabetEncoderTests.cs]: Verified and updated.

File: src/Tests.Marvin/CustomAlphabetEncoderTests.cs (hunk 2 above)
Comment (michaelcfanning) [reply src/Tests.Marvin/CustomAlphabetEncoderTests.cs]: This is just test code, so not sure the accessibility is very important. Many test frameworks requires a public method in order to invoke (I think XUnit does).

Speaking of which, why aren't we using XUnit? :) Can we do that? Or make the case for me to stick with MSTest.

:)

File: src/Tests.Marvin/CustomAlphabetEncoderTests.cs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+            encodeAction.Should().Throw<ArgumentNullException>();
+        }
+
+        /// <summary>
+        /// Test that the correct exception is thrown if string input is null on decode.
+        /// </summary>
+        [TestMethod]
+        public void CustomAlphabetEncoder_ShouldThrowExceptionWithNullInputOnDecodeByteArray()
+        {
+            CustomAlphabetEncoder testEncoder = new CustomAlphabetEncoder("abc");
```
Comment (michaelcfanning) [inline src/Tests.Marvin/CustomAlphabetEncoderTests.cs:129]: >CustomAlphabetEncoder [](http://example.com/codeflow?start=12&length=21)

our coding standard says you should use 'var' here because the type is explicit in the code (appears after 'new'). We do this because it makes source lines shorter and easier to read. The consistent use of var also helps lines up code at the left, again helping readability.

File: src/Tests.Marvin/CustomAlphabetEncoderTests.cs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+            string data = null;
+
+            Action decodeAction = () => testEncoder.Decode(data);
```
Comment (michaelcfanning) [inline src/Tests.Marvin/CustomAlphabetEncoderTests.cs:132]: >Action [](http://example.com/codeflow?start=12&length=6)

Here you would retain the explicit type, because it's not immediately clear from code that the return value is of your delegate (although most of us know this <g>).

File: src/Tests.Marvin/CustomAlphabetEncoderTests.cs (hunk 2 above)
Comment (eddynaka) [reply src/Tests.Marvin/CustomAlphabetEncoderTests.cs]: when we first created, we started as MSTest.
We didn't add xunit and a few other things we normally use (stylecop, etc.)

File: src/Tests.Marvin/CustomAlphabetEncoderTests.cs [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+            action.Should().Throw<ArgumentException>();
+        }
+
+        /// <summary>
+        /// Test that correct exception is thrown if byte array is out of the allowed uint range.
+        /// </summary>
+        /// <param name="alphabet">The alphabet with which to instantiate the CustomAlphabetEncoder.</param>
+        /// <param name="testInput">Example input which should throw an exception.</param>
+        [TestMethod]
+        [DataRow("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890.-_~", (long)UInt32.MinValue - 1)]
```
Comment (michaelcfanning) [inline src/Tests.Marvin/CustomAlphabetEncoderTests.cs:99]: >DataRow [](http://example.com/codeflow?start=9&length=7)

OK, so in general, the use of reflection in test frameworks holds significant potential to degrade performance. This was very true in the SARIF SDK (but this is a different framework, XUnit, and we noticed real slowdowns using the [Theory] mechanism.

This code pattern is very clean, though, so perhaps we can stick with it unless/until we observe that tests are slow. This is a small repo/package after all.

Based on what I know about the perf problems with the XUnit analog approach, maybe we should go ahead and stick with MSTest as well. :)

File: src/Tests.Marvin/CustomAlphabetEncoderTests.cs [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+using FluentAssertions;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Microsoft.Security.Utilities
+{
+
+    [TestClass]
+    public class CustomAlphabetEncoderTests
+    {
+        /// <summary>
```
Comment (michaelcfanning) [inline src/Tests.Marvin/CustomAlphabetEncoderTests.cs:14]: >summary [](http://example.com/codeflow?start=13&length=7)

OK, so something you and Eddy should being to think about is what coding conventions are appropriate for product code vs. test. XML docs are intended to provided IntelliSense for API consumers, and we don't expect anyone to consume test methods as an API (they are invoked by the framework).

And so, from my perspective, these XML doc comments aren't providing any value and they consume vertical space in the file.

Now, they may be helpful if it's hard to understand the purpose of each input arg, but in this case the parameter names are very clear.

This is just my opinion. I personally would not have authored these comments and I wouldn't be inclined to maintain them to keep the code consistent. Instead, I would focus on the clarity of my parameter names. If something needed to be explained further, I'd use a simple code comment.

Not that the case is very different for our actual API, the doc comments in this case will be consumed by arbitrary 3rd parties using our custom alphabet. *They* are the reason we take on the additional burden of authoring this sort of thing.

File: src/Marvin/CustomAlphabetEncoder.cs [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+{
+    /// <summary>
+    /// Custom encoder class.
+    /// </summary>
+    public class CustomAlphabetEncoder
+    {
+        internal const string DefaultBase62Alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+
+        [ThreadStatic]
+        private static StringBuilder sb;
```
Comment (michaelcfanning) [inline src/Marvin/CustomAlphabetEncoder.cs:19]: >StringBuilder [](http://example.com/codeflow?start=23&length=13)

Why have you only applied ThreadStatic to this variable and not the others?

It appears what you want to do (which is good!) is to reserve a single StringBuilder instance per thread for operatting against all encoder instances. I support this. :) But multiple encoder instances on the same thread could have a different alphabet, because you set it in the constructor. And so it appears that 'alphabet' should be an instance variable (as well as baseEncoding and charToValuerMap.

As you consider this, please make sure to prefix all static variables with 's_', whichever ones you retain. This makes it much easier to understand behavior when reading the remainder of the code.

Also, if you agree with me that there's a bug here, I suggest you first author the test that proves the failure exists. It appears this is as simple as:

1) Create one encoder instances with an alphabet and confirm it behaves properly, 
2) Create a second encoder instance with a *different* custom alphabet and confirm it behaves properly.
3) repeat your test with the first instance. I think it will now break, as you reset static data on the thread when you created the second encoder instance.

File: src/Marvin/CustomAlphabetEncoder.cs [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+        private static uint baseEncoding;
+        private static Dictionary<char, uint> charToValueMap;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CustomAlphabetEncoder"/> class.
+        /// </summary>
+        /// <param name="customAlphabet">The alphabet to be used for all encoding/decoding operations.</param>
+        public CustomAlphabetEncoder(string customAlphabet = DefaultBase62Alphabet)
+        {
+            alphabet = string.IsNullOrWhiteSpace(customAlphabet) ? DefaultBase62Alphabet : customAlphabet;
```
Comment (michaelcfanning) [inline src/Marvin/CustomAlphabetEncoder.cs:30]: >IsNullOrWhiteSpace [](http://example.com/codeflow?start=30&length=18)

so let's think about this. Someone went to the trouble of providing you "\t\r\n" as an alphabet, all whitespace (a tab, a return char, a newline). Is the right thing to do here to fallback to the defaultbase62? 

That seems non-obvious to me. I think I'd expect us to either honor this encoding alphabet (why not?) or to raise an exception (if in our opinion it is poor form to encode non-visible characters).

So what *is* the valid alphabet set? Latin? Alphanumeric plus some symbols?

File: src/Marvin/CustomAlphabetEncoder.cs [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+            baseEncoding = (uint)alphabet.Length;
+
+            charToValueMap = new Dictionary<char,uint>();
```
Comment (michaelcfanning) [inline src/Marvin/CustomAlphabetEncoder.cs:33]: >uint [](http://example.com/codeflow?start=49&length=4)

there should be a space preceding this.

File: src/Marvin/CustomAlphabetEncoder.cs [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+    /// Custom encoder class.
+    /// </summary>
+    public class CustomAlphabetEncoder
+    {
+        internal const string DefaultBase62Alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+
+        [ThreadStatic]
+        private static StringBuilder sb;
+        private static string alphabet;
+        private static uint baseEncoding;
```
Comment (michaelcfanning) [inline src/Marvin/CustomAlphabetEncoder.cs:21]: >baseEncoding [](http://example.com/codeflow?start=28&length=12)

You've named this 'baseEncoding' but it actually appears to be your count of alphabet characters. Should we use a different name?

Also, in the general spirit of keeping code as spare as possible, I would typically avoid declaring an instance variable for something that is very easy to retrieve or compute (like the length of the alphabet). i.e., consider whether it is sufficient in code to simply use `alphabet.Length` where this data is required. Just a minor thought.

File: src/Marvin/CustomAlphabetEncoder.cs [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+        /// Initializes a new instance of the <see cref="CustomAlphabetEncoder"/> class.
+        /// </summary>
+        /// <param name="customAlphabet">The alphabet to be used for all encoding/decoding operations.</param>
+        public CustomAlphabetEncoder(string customAlphabet = DefaultBase62Alphabet)
+        {
+            alphabet = string.IsNullOrWhiteSpace(customAlphabet) ? DefaultBase62Alphabet : customAlphabet;
+            baseEncoding = (uint)alphabet.Length;
+
+            charToValueMap = new Dictionary<char,uint>();
+            for (int i = 0; i < alphabet.Length; i++)
```
Comment (michaelcfanning) [inline src/Marvin/CustomAlphabetEncoder.cs:34]: >alphabet.Length [](http://example.com/codeflow?start=32&length=15)

Look! you used alphabet.Length here instead of baseEncoding. btw - I now see you appeared to do this in order to convert this value to uint and avoid various casts. Dunno, but I'd tend to avoid it. If you do retain it, use something like alphabetLength as a name?

File: src/Marvin/CustomAlphabetEncoder.cs [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+        /// </summary>
+        /// <param name="data">The byte array to encode. Must be 4 bytes or less.</param>
+        /// <returns>
+        /// The byte array encoded using the character set with which this class was instantiated.
+        /// </returns>
+        /// <exception cref="ArgumentNullException">Thrown if 'data' is null.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown if the length of the data array is greater than 4.</exception>
+        public string Encode(byte[] data)
+        {
+            if (data == null)
```
Comment (michaelcfanning) [inline src/Marvin/CustomAlphabetEncoder.cs:57]: >data [](http://example.com/codeflow?start=16&length=4)

Why can't we encode null? string.format will happily merge null into a string.

What does Convert.FromBase64String do? Maybe we could model ourselves after that API (when it is passed a null input, I mean).

File: src/Tests.Marvin/CustomAlphabetEncoderTests.cs (hunk 2 above)
Comment (michaelcfanning) [reply src/Tests.Marvin/CustomAlphabetEncoderTests.cs]: I am not aware that StyleCop is a team standard. :) My opinion is that we should be preferring the built-in VS formatting options (and not using StyleCop or CodeMaid)

Comment (eddynaka) [reply src/Tests.Marvin/CustomAlphabetEncoderTests.cs]: StyleCop is a nuget package where you can configure style things such as vars and many other rules and enforce during build.
This is a complementary to the .editorconfig

File: src/Marvin/CustomAlphabetEncoder.cs [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+            {
+                throw new ArgumentOutOfRangeException(nameof(data));
+            }
+
+            if (data.Length == 0)
+            {
+                return string.Empty;
+            }
+
+            uint convertedInput = BitConverter.ToUInt32(data, 0);
```
Comment (michaelcfanning) [inline src/Marvin/CustomAlphabetEncoder.cs:73]: >ToUInt32 [](http://example.com/codeflow?start=47&length=8)

Why did you make a decision to use the BitConverter class? Doesn't this helper only convert the bytes you point it to? And ignore the remainder of the buffer?

You can test this by encoding buffers of length 4 and 5, if they encode identically, that's a serious problem. What we want to do is encode the complete buffer. I provided some code to do this as an example, but it appears you didn't use that approach?

File: src/Tests.Marvin/CustomAlphabetEncoderTests.cs (hunk 2 above)
Comment (michaelcfanning) [reply src/Tests.Marvin/CustomAlphabetEncoderTests.cs]: I am definitely aware of what StyleCop is. I haven't participated in a conversation myself around what our formal standard is. :) Perhaps you or Mary could help us establish our recommendation. I am pretty sure whatever standard you have in mind isn't applied consistently across the code that we own.

File: src/Marvin/CustomAlphabetEncoder.cs (hunk 16 above)
Comment (michaelcfanning) [reply src/Marvin/CustomAlphabetEncoder.cs:73]: Btw - if you think I might be right about this, please go prove it by authoring the appropriate test cases that fail. :)

File: src/Tests.Marvin/CustomAlphabetEncoderTests.cs [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+
+using System;
+using FluentAssertions;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Microsoft.Security.Utilities
+{
+
+    [TestClass]
+    public class CustomAlphabetEncoderTests
```
Comment (michaelcfanning) [inline src/Tests.Marvin/CustomAlphabetEncoderTests.cs:12]: >CustomAlphabetEncoderTests [](http://example.com/codeflow?start=17&length=26)

Let describe a test:

Create an encoder using a custom alphabet.
Allocate a random set of bytes.
Encode the random bytes in the custom alphabet.
Verify the length of the encoded thing (you should be able to do this by understanding the relationship of the size of the input alphabet vs. encoding alphabet. In base64, for example, every 3 bytes result in 4 encoded characters).
Verify that the encoded value itself only contains legal characters.
[I note here, by the way, that your API doesn't include a facility for padding, which is typical in encoder implementations]
Now, take your encoded string and decode it, and it should match the input exactly.

You should have tests cover all corner cases, including circumstances where input bits perfectly align with encoding boundaries and all cases where they do not. In base64, for example, you would provoke all padding conditions, no padding, one character, or two. I would typically create tests that use a data length 'window' that doubles the test size. And start my tests from null, build them out one character at a time until the maximum is set.

File: src/Marvin/CustomAlphabetEncoder.cs (hunk 15 above)
Comment (marmegh) [reply src/Marvin/CustomAlphabetEncoder.cs:57]: Looking at the code for Convert.FromBase64String it throws an exception when passed a null value. The byte[] encoding of null is an empty array, which is handled for encoding and decoding.

Comment (michaelcfanning) [reply src/Marvin/CustomAlphabetEncoder.cs:57]: Got it. So, based on your feedback, I think I agree raising an exception on null is a good idea. Thanks for taking a look.

File: src/Marvin/CustomAlphabetEncoder.cs (hunk 16 above)
Comment (marmegh) [reply src/Marvin/CustomAlphabetEncoder.cs:73]: BitConvert.ToUint32 will convert the first four bytes of the byte array beginning at the indicated index to a uint.

{1, 2, 3, 4, 5}
{1, 2, 3, 4}
Both of the above yield the same result starting at index 0 (67305985).

However, we're currently enforcing that the byte array doesn't exceed a length of 4 to avoid this issue. A few lines above we throw an exception for this exact reason. Again, this is to match the current process used for the needed base62 encoded checksums, which relies on the uint32. If/when we add the additional base64 bitwise encoding logic, we will have a reason to reconsider this for any case using the alternate encodigng approach.

File: src/Tests.Marvin/CustomAlphabetEncoderTests.cs (hunk 8 above)
Comment (marmegh) [reply src/Tests.Marvin/CustomAlphabetEncoderTests.cs]: I tried both ways. There's less functionality with MSTest for the approach that was used in XUnit; but the data rows definitely make the tests run longer. I followed suit with other tests in the project.

File: src/Tests.Marvin/CustomAlphabetEncoderTests.cs [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+    {
+        [TestMethod]
+        public void CustomAlphabetEncoder_EncodeUintWithTestCases()
```
Comment (marmegh) [inline src/Tests.Marvin/CustomAlphabetEncoderTests.cs:19]: >CustomAlphabetEncoder_EncodeUintWithTestCases() [](http://example.com/codeflow?start=20&length=47)

Combined the test to encode byte[] input regardless of alphabet and input with default alphabet tests.

I compared the performance of those tests against using the testClass array and run time decreased by half.

File: src/Marvin/CustomAlphabetEncoder.cs [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+                throw new ArgumentNullException(nameof(encodedValue));
+            }
+
+            uint decodedValue = 0;
+
+            foreach (char c in encodedValue)
+            {
+                if(!charToValueMap.ContainsKey(c))
+                {
+                    throw new ArgumentException(nameof(encodedValue), "Alphabet does not contain all characters in input");
```
[the remark this answers is not in the document]
Comment (marmegh) [reply src/Marvin/CustomAlphabetEncoder.cs:98]: For the first part, addressed.

I agree on the need to aggregate validation errors. I'm aware of one good open source solution for addressing this. Adopting that is probably worth a discussion as a team.

File: src/Marvin/CustomAlphabetEncoder.cs [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+            {
+                // Repeated values in the custom alphabet will cause unreliable encoding/decoding.
+                if(charToValueMap.ContainsKey(alphabet[i]))
+                {
+                    throw new ArgumentException(nameof(customAlphabet), "Duplicate value detected in the alphabet.");
+                }
+
+                if (Char.IsWhiteSpace(alphabet[i]) || Char.IsSurrogate(alphabet[i]) || (int)alphabet[i] > 127)
+                {
+                    throw new ArgumentException(nameof(customAlphabet), String.Format("Forbidden character type detected in the alphabet: {0}.", alphabet[i]));
```
Comment (eddynaka) [inline src/Marvin/CustomAlphabetEncoder.cs:49]: >String.Format("Forbidden character type detected in the alphabet: {0}.", alphabet[i]) [](http://example.com/codeflow?start=72&length=85)

nit: replace string.format for $"text ... {alphabet[i]}"

File: src/Marvin/CustomAlphabetEncoder.cs [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+            if(alphabet.Length < 2)
+            {
+                throw new ArgumentException(nameof(customAlphabet), "Alphabet must be at least 2 characters.");
+            }
+
+            charToValueMap = new Dictionary<char, uint>();
+            for (int i = 0; i < alphabet.Length; i++)
+            {
+                // Repeated values in the custom alphabet will cause unreliable encoding/decoding.
+                if(charToValueMap.ContainsKey(alphabet[i]))
```
Comment (eddynaka) [inline src/Marvin/CustomAlphabetEncoder.cs:42]: >( [](http://example.com/codeflow?start=18&length=1)

fix the space before the parenthesis

File: src/Marvin/CustomAlphabetEncoder.cs [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+            if (encodedValue == null)
+            {
+                throw new ArgumentNullException(nameof(encodedValue));
+            }
+
+            uint decodedValue = 0;
+
+            foreach (char c in encodedValue)
+            {
+                if(!charToValueMap.ContainsKey(c))
```
Comment (eddynaka) [inline src/Marvin/CustomAlphabetEncoder.cs:96]: >( [](http://example.com/codeflow?start=18&length=1)

fix the space and add a new line after the }

File: src/Marvin/CustomAlphabetEncoder.cs [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+                {
+                    throw new ArgumentException(nameof(encodedValue), "Alphabet does not contain all characters in input");
+                }
+                decodedValue *= (uint)alphabet.Length;
```
Comment (eddynaka) [inline src/Marvin/CustomAlphabetEncoder.cs:100]: >(uint)alphabet.Length [](http://example.com/codeflow?start=32&length=21)

should this be calculated before the foreach to prevent being done on every iteration?

File: src/Marvin/CustomAlphabetEncoder.cs (hunk 20 above)
Comment (eddynaka) [reply src/Marvin/CustomAlphabetEncoder.cs]: the alphabet[i] should be inside {}.

File: src/Tests.Marvin/CustomAlphabetEncoderTests.cs [hunk 24]
```diff
@@ -0,0 +1,224 @@
...
+            }
+
+            return output;
+        }
+
+        private string GenerateAsciiString(int low = 0, int high = 256)
+        {
+            var sb = new StringBuilder();
+
+            for(int i = low; i < high; i++)
```
Comment (michaelcfanning) [inline src/Tests.Marvin/CustomAlphabetEncoderTests.cs:216]: >for [](http://example.com/codeflow?start=12&length=3)

UBER NIT: add a space here.

File: src/Marvin/CustomAlphabetEncoder.cs [hunk 25]
```diff
@@ -0,0 +1,109 @@
...
+        }
+
+        /// <summary>
+        /// Encode an unsigned integer (a checksum) in a given character set.
+        /// </summary>
+        /// <param name="data">The unsigned integer to encode.</param>
+        /// <returns>
+        /// The unsigned integer encoded using the character set with which this class was instantiated.
+        /// </returns>
+        public string Encode(uint data)
```
Comment (michaelcfanning) [inline src/Marvin/CustomAlphabetEncoder.cs:63]: >data [](http://example.com/codeflow?start=34&length=4)

so marvin32 comes in a 32-bit returning and 64-bit returning form, seems like we could provide 64-bit overloads/tests at not much cost.

**Review by michaelcfanning — APPROVED** [review]
:shipit:

**Review by michaelcfanning — APPROVED** [review]
:shipit:
